File: E/1807-evaluate-the-bracket-pairs-of-a-string/main.go
package main

import "fmt"

func evaluate(s string, knowledge [][]string) string {
	M := map[string]string{}
	for i := 0; i < len(knowledge); i++ {
		M[knowledge[i][0]] = knowledge[i][1]
	}

	ret := []byte{}
	key := []byte{}
	isIn := false
	for i := 0; i < len(s); i++ {
		if s[i] == byte('(') {
			isIn = true
			key = []byte{}
		} else if s[i] == byte(')') {
			isIn = false
			v, ok := M[string(key)]
			if !ok {
				ret = append(ret, '?')
			} else {
				for j := 0; j < len(v); j++ {
					ret = append(ret, v[j])
				}
			}
		} else if !isIn {
			ret = append(ret, s[i])
		} else {
			key = append(key, s[i])
		}
	}

	return string(ret)
}

func main() {
	fmt.Println()
}

File: E/1807-evaluate-the-bracket-pairs-of-a-string/main.rs
struct Solution {}

use std::collections::HashMap;
impl Solution {
  pub fn evaluate(s: String, knowledge: Vec<Vec<String>>) -> String {
    let mut is_in: bool = false;
    let mut buf: Vec<u8> = Vec::new();
    let mut ret: Vec<u8> = Vec::new();

    let mut m: HashMap<String, String> = knowledge
      .into_iter()
      .map(|x| (x[0].clone(), x[1].clone()))
      .collect();
    for b in s.as_bytes() {
      if *b == '(' as u8 {
        is_in = true;
        buf.clear();
      } else if *b == ')' as u8 {
        is_in = false;
        let k = String::from_utf8(buf.clone()).unwrap();
        if m.contains_key(&k) {
          ret.append(&mut m.get(&k).unwrap().as_bytes().to_vec());
        } else {
          ret.push('?' as u8);
        }
      } else {
        if is_in {
          buf.push(*b);
        } else {
          ret.push(*b);
        }
      }
    }
    String::from_utf8(ret).unwrap()
  }
}
